gno.land/pkg/gnoweb/markdown/ext_column.go

```diff
@@ -0,0 +1,313 @@
+	if node == nil {
+		return false
+	}
+
+	cnode, ok := node.(*ColumnNode)
+	return ok && cnode.Tag == tag
+}
+
+// parseLineTag checks if the line starts with any of the tag.
+func parseLineTag(ctx *columnContext, line []byte) ColumnTag {
+	line = util.TrimRightSpace(util.TrimLeftSpace(line))
+
+	switch string(line) {
+	case "": // empty line
+		return ColumnTagUndefined
+	case "<gno-columns>":
+		return ColumnTagOpen
+	case "</gno-columns>":
+		return ColumnTagClose
+	case ":::":
+		if ctx.IsOpen {
+			return ColumnTagClose
+		}
+		return ColumnTagOpen
+	}
+
+	if line[0] == '#' {
+		return ColumnTagSep
+	}
+
+	return ColumnTagUndefined
+}
+
+// Open create a column node based on line tag.
+func (p *columnParser) Open(self ast.Node, reader text.Reader, pc parser.Context) (ast.Node, parser.State) {
+	const MaxHeading = 6
+
+	// Columns tag cannot be a child of another node
+	if self.Parent() != nil {
+		return nil, parser.NoChildren
+	}
+
+	// Get column context
+	cctx, ok := pc.Get(columnContextKey).(*columnContext)
+	if !ok || !cctx.IsOpen {
+		cctx = &columnContext{} // new context
+		pc.Set(columnContextKey, cctx)
+	}
+
+	line, segment := reader.PeekLine()
+
+	tagKind := parseLineTag(cctx, line)
+	if tagKind == ColumnTagUndefined {
+		return nil, parser.NoChildren
+	}
+
+	node := NewColumn(cctx, cctx.Index, tagKind)
+	switch tagKind {
+	case ColumnTagSep:
+		if !cctx.IsOpen {
+			return nil, parser.NoChildren
+		}
+
+		level, maxLevel := 1, min(len(line), MaxHeading+1)
+		for level < maxLevel && line[level] == '#' {
+			level++
+		}
+
+		switch {
+		case level >= MaxHeading:
+			// Level is beyond the maximum one, ignore this heading
+			return nil, parser.NoChildren
+		case cctx.RefHeadingLevel == 0:
+			// Register first header as reference
+			cctx.RefHeadingLevel = level
+		case cctx.RefHeadingLevel != level:
+			// If heading level reference is different, skip it
+			return nil, parser.NoChildren
+		}
+
+		// Process creating a column
+		cctx.Index++
+		node.Index = cctx.Index
+
+		if trimmed := util.TrimLeft(line[level:], []byte{' ', '\n'}); len(trimmed) == 0 {
+			// Empty heading, create a column separator and skip the parsing
+			reader.Advance(segment.Len())
+			return node, parser.NoChildren
+		}
+
+		// Insert a column separator but return an empty node so we can
+		// let the parser parse the heading
+		self.InsertBefore(self, self.PreviousSibling(), node)
+
+	case ColumnTagOpen:
+		if !cctx.IsOpen {
+			cctx.IsOpen = true
+			return node, parser.NoChildren
+		}
+
+	case ColumnTagClose:
+		if cctx.IsOpen {
+			cctx.IsOpen = false
+			return node, parser.NoChildren
+		}
+	}
+
+	// Ignore node
+	return nil, parser.NoChildren
+}
+
+func (*columnParser) Continue(n ast.Node, reader text.Reader, _ parser.Context) parser.State {
+	return parser.Close
+}
+
+func (*columnParser) Close(_ ast.Node, reader text.Reader, _ parser.Context) {}
+
+func (*columnParser) CanInterruptParagraph() bool {
```

Comment:
deadcode?

Reply:
Those methods are for block parsing interface; I will add an assert reference.